File: datasets/C++/Mistral/Algorithm/hard4.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

vector<int> reverseFactorization(int N, vector<int> A) {
    sort(A.begin(), A.end());
    queue<pair<int, vector<int>>> q;
    q.push({1, {}});

    while (!q.empty()) {
        auto current = q.front();
        q.pop();
        int currentValue = current.first;
        vector<int> sequence = current.second;

        if (currentValue == N) {
            return sequence;
        }

        for (int a : A) {
            long long nextValue = static_cast<long long>(currentValue) * a;
            if (nextValue <= N) {
                vector<int> newSequence = sequence;
                newSequence.push_back(nextValue);
                q.push({static_cast<int>(nextValue), newSequence});
            }
        }
    }

    return {-1};
}

int main() {
    int N, K;
    cout << "Enter N and K: ";
    cin >> N >> K;

    vector<int> A(K);
    cout << "Enter the elements of A: ";
    for (int i = 0; i < K; ++i) {
        cin >> A[i];
    }

    vector<int> result = reverseFactorization(N, A);
    if (result.size() == 1 && result[0] == -1) {
        cout << -1 << endl;
    } else {
        cout << "1 ";
        for (int num : result) {
            cout << num << " ";
        }
        cout << endl;
    }

    return 0;
}

File: datasets/Python/Mistral/Algorithm/hard4.py
from collections import deque

def reverse_factorization(N, A):
    A.sort()
    queue = deque()
    queue.append((1, []))  # (current value, sequence)

    while queue:
        current_value, sequence = queue.popleft()

        if current_value == N:
            return sequence

        for a in A:
            next_value = current_value * a
            if next_value <= N:
                queue.append((next_value, sequence + [next_value]))

    return [-1]

# Input
N, K = map(int, input("Enter N and K: ").split())
A = list(map(int, input("Enter the elements of A: ").split()))

# Calculate and print the result
result = reverse_factorization(N, A)
if result == [-1]:
    print(-1)
else:
    print(' '.join(map(str, [1] + result)))
